### PR TITLE
fix(metrics): Delete StringIndexer and PerfStringIndexer past TTL

### PR DESCRIPTION
**Pull request by john-z-yang**
### Overview
Adding cleanup tasks for `StringIndexer` and `PerfStringIndexer`. This makes sure we delete rows after 90 days to stay compliant.

### Changes
- Add `StringIndexer` and `PrefStringIndexer` to `BULK_QUERY_DELETES`
- Update [ops/.../services/getsentry/_values.yaml](https://github.com/getsentry/ops/blob/master/k8s/services/getsentry/_values.yaml#L283) to supply additional `PerfStringIndexer` to `--model` command line arguments (see https://github.com/getsentry/sentry/pull/45673)
- Update [ops/.../services/getsentry/_values.yaml](https://github.com/getsentry/ops/blob/master/k8s/services/getsentry/_values.yaml#L283) to supply additional `StringIndexer` to `--model` command line arguments (see https://github.com/getsentry/ops/pull/6301)

Closes [SNS-2109](https://getsentry.atlassian.net/browse/SNS-2109)

[SNS-2109]: https://getsentry.atlassian.net/browse/SNS-2109?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/sentry/runner/commands/cleanup.py
Comment (wedamija) [inline src/sentry/runner/commands/cleanup.py:187]: Ahh, it looks like we conditionally use postgres as part of the indexer based on `SENTRY_METRICS_INDEXER`.

So I think that the way you should add these is to add them to `EXTRA_BULK_QUERY_DELETES` programatically when we instantiate https://github.com/getsentry/sentry/blob/0c225ed01bd03f92ff4925cfa019b551fab9684f/src/sentry/sentry_metrics/indexer/postgres/postgres_v2.py#L212-L214

That way, if we configure a different indexer we'll stop attempting to clean this up.

Comment (john-z-yang) [reply src/sentry/runner/commands/cleanup.py:187]: Thanks, by programmatically I'm assuming you mean something like
```python
def __init__(self) -> None:
        super().__init__(CachingIndexer(indexer_cache, PGStringIndexerV2()))
        cleanup.EXTRA_BULK_QUERY_DELETES.extend(
            [
                (StringIndexer, "last_seen", None),
                (PerfStringIndexer, "last_seen", None),
            ]
        )
```

**Review by wedamija — APPROVED** [review]

**Comment by onewland** [comment]
Is there a limiter on this (how many rows get deleted per run?) or are we holding off on merge until we delete the huge backlog on both tables in prod?

**Comment by wedamija** [comment]
> Is there a limiter on this (how many rows get deleted per run?) or are we holding off on merge until we delete the huge backlog on both tables in prod?

There's no limiter. @beezz can probably help us shepherd the first wave of deletions here.

File: src/sentry/sentry_metrics/indexer/postgres/postgres_v2.py
Comment (beezz) [inline src/sentry/sentry_metrics/indexer/postgres/postgres_v2.py:227]: This does not work. Where did you plan to instantiate the `PostgresIndexer` for the changes to `EXTRA_BULK_QUERY_DELETES` to take effect?

```
❯ git rev-parse --abbrev-ref HEAD
john/metrics_indexer_cleanup

❯ git --no-pager diff
diff --git a/src/sentry/runner/commands/cleanup.py b/src/sentry/runner/commands/cleanup.py
index 5e1c180883..b7273476e2 100644
--- a/src/sentry/runner/commands/cleanup.py
+++ b/src/sentry/runner/commands/cleanup.py
@@ -174,6 +174,10 @@ def cleanup(days, project, concurrency, silent, model, router, timed):
                 return False
             return model.__name__.lower() not in model_list

+
+        print(f"{EXTRA_BULK_QUERY_DELETES=}")
+        exit(1)
+
         # Deletions that use `BulkDeleteQuery` (and don't need to worry about child relations)
         # (model, datetime_field, order_by)
         BULK_QUERY_DELETES = [

❯ sentry cleanup
{"name":"sentry.utils.geo","event":"settings.GEOIP_PATH_MMDB not configured.","level":"warning"}
{"name":"sentry.plugins.github","event":"apps-not-configured","level":"info"}
EXTRA_BULK_QUERY_DELETES=[]
```

I understand that we might not use the `PostgresIndexer` but since the models are present always, worst case, we query empty tables, so I suggest adding the models to the `BULK_QUERY_DELETES`.

**Review by beezz — CHANGES_REQUESTED** [review]

**Comment by beezz** [comment]
> > Is there a limiter on this (how many rows get deleted per run?) or are we holding off on merge until we delete the huge backlog on both tables in prod?
> 
> There's no limiter. @beezz can probably help us shepherd the first wave of deletions here.

This part can be merged and if there's a large backlog to delete, the first run is done in a controlled way to delete the backlog and then schedule task with enough concurrency to keep up.

File: src/sentry/sentry_metrics/indexer/postgres/postgres_v2.py
Comment (john-z-yang) [reply src/sentry/sentry_metrics/indexer/postgres/postgres_v2.py:227]: > This does not work. Where did you plan to instantiate the PostgresIndexer for the changes to `EXTRA_BULK_QUERY_DELETES` to take effect?
I was plagiarizing from [6327](https://github.com/getsentry/sentry/pull/6327/files#diff-7173069809318e0fe8909c21eac05d4cef36e20cd1e14d7833d3bd135bd95625R64).
From @wedamija's comment I was expecting `PostgresIndexer` to be [loaded](https://github.com/getsentry/sentry/blob/00845f9617523b5a98f07b2457696f6ebca8a33b/src/sentry/sentry_metrics/consumers/indexer/processing.py#L24) and [instantiated](https://github.com/getsentry/sentry/blob/00845f9617523b5a98f07b2457696f6ebca8a33b/src/sentry/sentry_metrics/consumers/indexer/processing.py#L31) in `processing.py`, which will update `EXTRA_BULK_QUERY_DELETES`. But I could be wrong because this is my first time working with the sentry codebase :)

Comment (wedamija) [reply src/sentry/sentry_metrics/indexer/postgres/postgres_v2.py:227]: Might be easier to just assume it exists and go back to how we implemented it earlier, since I guess there's a chance that the code doesn't end up imported

**Comment by onewland** [comment]
Do we know how many rows expire per day? Does this cron do batching for them automatically? 

It might be fine but if we're deleting 100k rows/day I could see that affecting Postgres performance (that's about how many new rows got written in prod in the last 24h)

**Comment by wedamija** [comment]
> Do we know how many rows expire per day? Does this cron do batching for them automatically?
> 
> It might be fine but if we're deleting 100k rows/day I could see that affecting Postgres performance (that's about how many new rows got written in prod in the last 24h) if it happens in a single statement.

It's definitely not deleting 100k rows in a single query. It's using our bulk deletion functionality here https://github.com/getsentry/sentry/blob/2ec9adaa218e4097caa36e456c160b7cbe906b94/src/sentry/runner/commands/cleanup.py#L272-L295. Batch size is 10k, and this is used for a fair few other high volume postgres tables.

**Comment by onewland** [comment]
Cool, that works for me.

File: src/sentry/sentry_metrics/indexer/postgres/postgres_v2.py
Comment (john-z-yang) [reply src/sentry/sentry_metrics/indexer/postgres/postgres_v2.py]: Sounds good, reverting to original

**Review by beezz — APPROVED** [review]

**Pull request merged**
